File: frontend/lib/videos.ts
import * as FileSystem from 'expo-file-system/legacy';
import { api } from './api';

export interface Video {
  videoId: string;
  userId: string;
  exercise: string;
  notes: string;
  sessionDate: string;
  uploadedAt: string;
  uploaded: boolean;
  s3Key: string;
  contentType: string;
  playbackUrl: string | null;
  durationSec?: number;
}

export interface PostVideoResponse {
  videoId: string;
  uploadUrl: string;
  expiresIn: number;
}

export async function listVideos(): Promise<Video[]> {
  const resp = await api.get<{ items: Video[] }>('/videos');
  return resp.items;
}

export async function fetchVideo(videoId: string): Promise<Video> {
  return api.get<Video>(`/videos/${videoId}`);
}

export async function createVideoRecord(params: {
  exercise: string;
  notes: string;
  sessionDate: string;
  contentType: string;
}): Promise<PostVideoResponse> {
  return api.post<PostVideoResponse>('/videos', params);
}

export async function uploadVideoFile(
  uploadUrl: string,
  fileUri: string,
  contentType: string,
  onProgress: (fraction: number) => void,
): Promise<void> {
  const task = FileSystem.createUploadTask(
    uploadUrl,
    fileUri,
    {
      httpMethod: 'PUT',
      uploadType: FileSystem.FileSystemUploadType.BINARY_CONTENT,
      headers: { 'Content-Type': contentType },
    },
    ({ totalBytesSent, totalBytesExpectedToSend }) => {
      if (totalBytesExpectedToSend > 0) {
        onProgress(totalBytesSent / totalBytesExpectedToSend);
      }
    },
  );
  const result = await task.uploadAsync();
  if (result && result.status >= 400) {
    throw new Error(`Upload failed: HTTP ${result.status}`);
  }
}

File: backend/api/videos.py
import json
import os
import time
from datetime import datetime, timezone
from decimal import Decimal

from boto3.dynamodb.conditions import Attr, Key

from api.auth import current_user
from api.db import videos_table
from api.storage import presigned_put, signed_playback_url


def _video_id() -> str:
    ts = int(time.time() * 1000)
    rand = os.urandom(6).hex()
    return f"v_{ts:013x}{rand}"


def _serialize(item: dict) -> dict:
    """Convert DynamoDB Decimal values to int/float for JSON serialisation."""
    out = {}
    for k, v in item.items():
        if isinstance(v, Decimal):
            out[k] = int(v) if v == v.to_integral_value() else float(v)
        else:
            out[k] = v
    return out


def _add_playback_url(item: dict) -> dict:
    if item.get("uploaded") and item.get("s3Key"):
        item["playbackUrl"] = signed_playback_url(item["s3Key"])
    else:
        item["playbackUrl"] = None
    return item


def post_videos(event: dict, context) -> dict:
    user = current_user(event)
    body = json.loads(event.get("body") or "{}")

    exercise = body.get("exercise", "").strip()
    notes = body.get("notes", "").strip()
    session_date = body.get("sessionDate") or datetime.now(timezone.utc).date().isoformat()
    content_type = body.get("contentType", "video/quicktime")

    video_id = _video_id()
    s3_key = f"videos/{user.sub}/{video_id}.mov"
    uploaded_at = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")

    videos_table().put_item(Item={
        "userId": user.sub,
        "videoId": video_id,
        "gsiPk": "all",
        "exercise": exercise,
        "notes": notes,
        "sessionDate": session_date,
        "uploadedAt": uploaded_at,
        "uploaded": False,
        "s3Key": s3_key,
        "contentType": content_type,
    })

    upload_url = presigned_put(s3_key, content_type)

    return {
        "statusCode": 201,
        "headers": {"Content-Type": "application/json"},
        "body": json.dumps({
            "videoId": video_id,
            "uploadUrl": upload_url,
            "expiresIn": 900,
        }),
    }


def get_videos(event: dict, context) -> dict:
    user = current_user(event)

    if user.is_coach():
        # All videos sorted by date descending via the byDate GSI.
        resp = videos_table().query(
            IndexName="byDate",
            KeyConditionExpression=Key("gsiPk").eq("all"),
            ScanIndexForward=False,
        )
    else:
        # Athlete's own videos, newest first.
        resp = videos_table().query(
            KeyConditionExpression=Key("userId").eq(user.sub),
            ScanIndexForward=False,
        )

    items = [_add_playback_url(_serialize(item)) for item in resp.get("Items", [])]

    return {
        "statusCode": 200,
        "headers": {"Content-Type": "application/json"},
        "body": json.dumps({"items": items}),
    }


def get_video(event: dict, context) -> dict:
    user = current_user(event)
    video_id = event["pathParameters"]["videoId"]

    if user.is_coach():
        # Look up by videoId across the GSI — fine at MVP scale (<100 videos).
        resp = videos_table().query(
            IndexName="byDate",
            KeyConditionExpression=Key("gsiPk").eq("all"),
            FilterExpression=Attr("videoId").eq(video_id),
        )
        items = resp.get("Items", [])
        if not items:
            return {"statusCode": 404, "headers": {"Content-Type": "application/json"},
                    "body": json.dumps({"error": "not found"})}
        item = items[0]
    else:
        resp = videos_table().get_item(Key={"userId": user.sub, "videoId": video_id})
        item = resp.get("Item")
        if not item:
            return {"statusCode": 404, "headers": {"Content-Type": "application/json"},
                    "body": json.dumps({"error": "not found"})}

    item = _add_playback_url(_serialize(item))

    return {
        "statusCode": 200,
        "headers": {"Content-Type": "application/json"},
        "body": json.dumps(item),
    }
